File: packages/shared-ts/src/chat/attachments.ts
import { IMAGE_GENERATION_MODEL_ID, VIDEO_GENERATION_MODEL_ID } from './routing';

const TEXT_ATTACHMENT_ACCEPT = ['text/*', 'application/json', 'application/xml'] as const;

const IMAGE_ATTACHMENT_ACCEPT = ['image/jpeg', 'image/png', 'image/gif', 'image/webp'] as const;

const AUDIO_ATTACHMENT_ACCEPT = [
  'audio/wav',
  'audio/mpeg',
  'audio/mp3',
  'audio/webm',
  'audio/ogg',
] as const;

const VIDEO_ATTACHMENT_ACCEPT = ['video/mp4', 'video/webm'] as const;

const OFFICE_ATTACHMENT_ACCEPT = [
  'application/pdf',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'application/vnd.openxmlformats-officedocument.presentationml.presentation',
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  'text/csv',
] as const;

export const DEFAULT_PROMPT_ATTACHMENT_ACCEPT = [
  ...TEXT_ATTACHMENT_ACCEPT,
  ...IMAGE_ATTACHMENT_ACCEPT,
  ...AUDIO_ATTACHMENT_ACCEPT,
].join(',');

export const OPENAI_PROMPT_ATTACHMENT_ACCEPT = [
  DEFAULT_PROMPT_ATTACHMENT_ACCEPT,
  ...OFFICE_ATTACHMENT_ACCEPT,
].join(',');

export const GEMINI_PROMPT_ATTACHMENT_ACCEPT = [
  DEFAULT_PROMPT_ATTACHMENT_ACCEPT,
  ...VIDEO_ATTACHMENT_ACCEPT,
  ...OFFICE_ATTACHMENT_ACCEPT,
].join(',');

export const VIDEO_GENERATION_PROMPT_ATTACHMENT_ACCEPT = [
  DEFAULT_PROMPT_ATTACHMENT_ACCEPT,
  ...VIDEO_ATTACHMENT_ACCEPT,
].join(',');

export interface PromptSelectableModelOption {
  id: string;
}

export const isGeminiModelId = (modelId?: string | null): boolean =>
  Boolean(modelId?.toLowerCase().includes('gemini'));

export const isVideoGenerationModelId = (modelId?: string | null): boolean =>
  modelId?.toLowerCase() === VIDEO_GENERATION_MODEL_ID;

export const isOpenAIModelId = (modelId?: string | null): boolean => {
  const normalizedModelId = modelId?.toLowerCase();
  if (!normalizedModelId) {
    return false;
  }

  return (
    normalizedModelId.includes('gpt') ||
    normalizedModelId.includes('o1') ||
    normalizedModelId.includes('o3')
  );
};

export const buildPromptAttachmentAccept = (modelId?: string | null): string => {
  if (isVideoGenerationModelId(modelId)) {
    return VIDEO_GENERATION_PROMPT_ATTACHMENT_ACCEPT;
  }

  if (isGeminiModelId(modelId)) {
    return GEMINI_PROMPT_ATTACHMENT_ACCEPT;
  }

  if (isOpenAIModelId(modelId)) {
    return OPENAI_PROMPT_ATTACHMENT_ACCEPT;
  }

  return DEFAULT_PROMPT_ATTACHMENT_ACCEPT;
};

export const filterPromptSelectableModelOptions = <T extends PromptSelectableModelOption>(
  modelOptions: readonly T[]
): T[] =>
  modelOptions.filter(
    (option) => option.id !== IMAGE_GENERATION_MODEL_ID && option.id !== VIDEO_GENERATION_MODEL_ID
  );

File: packages/core/go/pkg/engine/attachments.go
package engine

import "strings"

const (
	MaxAttachments             = 5
	MaxAttachmentBytes         = 20 * 1024 * 1024  // 20 MB decoded for standard, image, audio, and document attachments.
	MaxVideoAttachmentBytes    = 100 * 1024 * 1024 // 100 MB decoded for video attachments.
	MaxTotalAttachmentBytes    = 100 * 1024 * 1024 // 100 MB total decoded payload across all attachments.
	ImageGenerationModelID     = "google/gemini-2.5-flash-image-preview"
	VideoGenerationModelID     = "xai/grok-imagine-video-1.5"
	imageGenerationModelMarker = "gemini-2.5-flash-image"
)

var supportedImageMIME = map[string]bool{
	"image/jpeg": true,
	"image/png":  true,
	"image/gif":  true,
	"image/webp": true,
}

var supportedVideoMIME = map[string]bool{
	"video/mp4":  true,
	"video/webm": true,
}

var supportedDocumentMIME = map[string]bool{
	"application/pdf": true,
	"application/vnd.openxmlformats-officedocument.wordprocessingml.document":   true,
	"application/vnd.openxmlformats-officedocument.presentationml.presentation": true,
	"text/plain":       true,
	"text/markdown":    true,
	"application/json": true,
}

var supportedSpreadsheetMIME = map[string]bool{
	"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": true,
	"text/csv": true,
}

var officeMIMEByExtension = map[string]string{
	".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
	".pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
	".xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
}

func NormalizeAttachmentMIME(value string) string {
	trimmed := strings.TrimSpace(value)
	if idx := strings.Index(trimmed, ";"); idx != -1 {
		trimmed = strings.TrimSpace(trimmed[:idx])
	}
	return strings.ToLower(trimmed)
}

func NormalizeUploadedAttachmentMIME(filename, detectedMIME string) string {
	normalized := NormalizeAttachmentMIME(detectedMIME)
	ext := strings.ToLower(strings.TrimSpace(filenameExtension(filename)))

	if mapped, ok := officeMIMEByExtension[ext]; ok {
		if normalized == "" || normalized == "application/zip" || normalized == "application/octet-stream" {
			return mapped
		}
	}

	if ext == ".csv" && (normalized == "text/plain" || normalized == "application/octet-stream") {
		return "text/csv"
	}

	return normalized
}

func IsSupportedAudioAttachmentMIME(mime string) bool {
	mime = NormalizeAttachmentMIME(mime)
	if !strings.HasPrefix(mime, "audio/") {
		return false
	}
	return strings.HasSuffix(mime, "/wav") || strings.HasSuffix(mime, "/wave") ||
		strings.HasSuffix(mime, "/mpeg") || strings.HasSuffix(mime, "/mp3") ||
		strings.HasSuffix(mime, "/webm") || strings.HasSuffix(mime, "/ogg")
}

func IsSupportedDocumentAttachmentMIME(mime string) bool {
	mime = NormalizeAttachmentMIME(mime)
	return supportedDocumentMIME[mime] || supportedSpreadsheetMIME[mime]
}

func IsVideoAttachmentMIME(mime string) bool {
	return strings.HasPrefix(NormalizeAttachmentMIME(mime), "video/")
}

func ModelSupportsVideoAttachments(modelID string) bool {
	normalized := strings.ToLower(strings.TrimSpace(modelID))
	return strings.Contains(normalized, "gemini")
}

func IsMediaGenerationModelID(modelID string) bool {
	return IsImageGenerationModelID(modelID) || strings.TrimSpace(strings.ToLower(modelID)) == VideoGenerationModelID
}

func IsImageGenerationModelID(modelID string) bool {
	model := strings.TrimSpace(strings.ToLower(modelID))
	return strings.Contains(model, imageGenerationModelMarker)
}

func NormalizeImageGenerationModelID(modelID string) string {
	trimmed := strings.TrimSpace(modelID)
	if strings.EqualFold(trimmed, "google/gemini-2.5-flash-image") ||
		strings.EqualFold(trimmed, "gemini-2.5-flash-image") {
		return ImageGenerationModelID
	}
	return trimmed
}

func AttachmentByteLimit(mime string) (int64, bool, string) {
	normalized := NormalizeAttachmentMIME(mime)
	switch {
	case strings.HasPrefix(normalized, "video/"):
		return attachmentLimitResult(MaxVideoAttachmentBytes, supportedVideoMIME[normalized], `unsupported video type "%s"; allowed: video/mp4, video/webm`)
	case strings.HasPrefix(normalized, "audio/"):
		return attachmentLimitResult(MaxAttachmentBytes, IsSupportedAudioAttachmentMIME(normalized), `unsupported audio format "%s"; allowed: wav, mp3, webm, ogg`)
	case strings.HasPrefix(normalized, "image/"):
		return attachmentLimitResult(MaxAttachmentBytes, supportedImageMIME[normalized], `unsupported image type "%s"; allowed: image/jpeg, image/png, image/gif, image/webp`)
	case IsSupportedDocumentAttachmentMIME(normalized):
		return MaxAttachmentBytes, true, ""
	default:
		return MaxAttachmentBytes, false, `unsupported attachment type "%s"`
	}
}

func attachmentLimitResult(limit int64, supported bool, unsupportedTemplate string) (int64, bool, string) {
	if supported {
		return limit, true, ""
	}
	return limit, false, unsupportedTemplate
}

func filenameExtension(filename string) string {
	trimmed := strings.TrimSpace(filename)
	if trimmed == "" {
		return ""
	}
	slash := strings.LastIndexAny(trimmed, `/\`)
	if slash >= 0 {
		trimmed = trimmed[slash+1:]
	}
	dot := strings.LastIndex(trimmed, ".")
	if dot <= 0 || dot == len(trimmed)-1 {
		return ""
	}
	return trimmed[dot:]
}
